src/3d/symbols/qgspointcloud3dsymbol_p.cpp

```diff
@@ -227,6 +267,179 @@ void QgsPointCloud3DSymbolHandler::makeEntity( Qt3DCore::QEntity *parent, const
+    outNormal.normals.resize( ( outNormal.positions.count() ) * sizeof( float ) * 3 );
+    float *normPtr = reinterpret_cast<float *>( outNormal.normals.data() );
+    for ( int i = 0; i < normals.size(); ++i )
+    {
+      QVector3D normal = normals.at( i );
+      normal = normal.normalized();
+
+      *normPtr++ = normal.x();
+      *normPtr++ = normal.y();
+      *normPtr++ = normal.z();
+    }
+
+    // We apply a filter on triangles to improve the rendering:
+    // - we keep only triangles that have a least one point in the bounding box
+    // - if option is selected, we don't keep triangle with side greater than a threshold
+
+    outNormal.triangles.resize( triangleIndexes.size() * sizeof( quint32 ) );
+    quint32 *indexPtr = reinterpret_cast<quint32 *>( outNormal.triangles.data() );
+    size_t effective = 0;
+
+    bool sizeFilter = context.symbol()->filterTrianglesBySize();
+    bool heightFilter = context.symbol()->filterTrianglesByHeight();
+    float sizeThreshold =  context.symbol()->triangleSizeThreshold();
+    float heightThreshold =  context.symbol()->triangleHeightThreshold();
+
+    for ( size_t i = 0; i < triangleIndexes.size() / 3; ++i )
+    {
+      bool atLeastOneInBox = false;
+      bool greaterThanSize = false;
+      bool greaterThanHeight = false;
+      for ( size_t j = 0; j < 3; j++ )
+      {
+        QVector3D pos = outNormal.positions.at( triangleIndexes.at( i * 3 + j ) );
+        QVector3D pos2;
+
+        atLeastOneInBox |= bbox.intersects( pos.x(), pos.y(), pos.z() );
+
+        if ( heightFilter || sizeFilter )
+          pos2 = outNormal.positions.at( triangleIndexes.at( i * 3 + ( j + 1 ) % 3 ) );
+
+        if ( heightFilter )
+          greaterThanHeight |= std::fabs( pos.y() - pos2.y() ) > heightThreshold;
+
+        if ( sizeFilter && ! greaterThanHeight )
+        {
+          QVector3D pos2 = outNormal.positions.at( triangleIndexes.at( i * 3 + ( j + 1 ) % 3 ) );
+          // filter only in the horizontal plan, it is a 2.5D triangulation.
+          pos2.setY( 0 );
+          pos.setY( 0 );
+          greaterThanSize |= pos2.distanceToPoint( pos ) > sizeThreshold;
+        }
+
+        if ( greaterThanSize || greaterThanHeight )
+          break;
+      }
+      if ( atLeastOneInBox && !greaterThanSize && !greaterThanHeight )
+      {
+        for ( size_t j = 0; j < 3; j++ )
+        {
+          size_t vertIndex = triangleIndexes.at( i * 3 + j );
+          *indexPtr++ = quint32( vertIndex );
+        }
+        effective++;
+      }
+    }
+
+    if ( effective != 0 )
+    {
+      outNormal.triangles.resize( effective * 3 * sizeof( quint32 ) );
+    }
+    else
+    {
+      outNormal.triangles.clear();
+      outNormal.normals.clear();
+    }
+  }
+  catch ( std::exception &e )
+  {
+    // something went wrong, better to retrieve initial state
+    QgsDebugMsgLevel( QStringLiteral( "Error with triangulation" ), 4 );
+    outNormal = PointData();
+    processNode( pc, n, context );
```

Comment:
yes, it is. If we got an exception of Delaunator, we give up and return to point cloud. I've got some exceptions thrown by Delaunator during testing implementation, mainly due to bad implementation, but I'd prefer keep the try/catch to prevent any fail due to any bad points that could appears.

Reply:
sounds good, I am mainly interested if it is normal to get exceptions and under what circumstances - and whether we can avoid them